paimon-common/src/main/java/org/apache/paimon/utils/InternalRowUtils.java

```diff
@@ -48,13 +48,74 @@
 import java.math.BigDecimal;
 import java.math.RoundingMode;
 import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 
 /** Utils for {@link InternalRow} structures. */
 public class InternalRowUtils {
 
+    public static boolean equals(Object data1, Object data2, DataType dataType) {
+        if ((data1 == null) != (data2 == null)) {
+            return false;
+        }
+        if (data1 != null) {
+            if (data1 instanceof InternalRow) {
```

Comment:
@JingsongLi  both array and map have their own equals and hashcode, so I leave it to use o.equals

Reply:
But they are not safe, if there is a row inside array or map, it is not safe too.

So it is better to add implementation for array and map too.

Reply:
addressed